src/Transport/Receiving/MessagePump.cs

```diff
@@ -0,0 +1,295 @@
+
+                            scope?.Complete();
+                        }
+                    }
+
+                    if (receiveCancellationTokenSource.IsCancellationRequested)
+                    {
+                        await receiver.SafeAbandonAsync(pushSettings.RequiredTransactionMode, lockToken).ConfigureAwait(false);
+                    }
+                }
+            }
+            catch (Exception exception)
+            {
+                try
+                {
+                    ErrorHandleResult result;
+
+                    using (var scope = CreateTransactionScope())
+                    {
+                        var transportTransaction = CreateTransportTransaction(message.PartitionKey);
+
+                        var errorContext = new ErrorContext(exception, message.GetNServiceBusHeaders(), messageId, body, transportTransaction, message.SystemProperties.DeliveryCount);
+
+                        result = await onError(errorContext).ConfigureAwait(false);
+
+                        if (result == ErrorHandleResult.Handled)
+                        {
+                            await receiver.SafeCompleteAsync(pushSettings.RequiredTransactionMode, lockToken).ConfigureAwait(false);
+                        }
+
+                        scope?.Complete();
+                    }
+
+                    if (result == ErrorHandleResult.RetryRequired)
+                    {
+                        await receiver.SafeAbandonAsync(pushSettings.RequiredTransactionMode, lockToken).ConfigureAwait(false);
+                    }
+                }
+                catch (Exception onErrorException)
+                {
+                    logger.WarnFormat("Recoverability failed for message with ID {0}. The message will be retried. Exception details: {1}", messageId, onErrorException);
+
+                    await receiver.SafeAbandonAsync(pushSettings.RequiredTransactionMode, lockToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        TransactionScope CreateTransactionScope()
+        {
+            return pushSettings.RequiredTransactionMode == TransportTransactionMode.SendsAtomicWithReceive
+                ? new TransactionScope(TransactionScopeOption.RequiresNew, new TransactionOptions
+                {
+                    IsolationLevel = IsolationLevel.Serializable,
+                    Timeout = TransactionManager.MaximumTimeout
+                }, TransactionScopeAsyncFlowOption.Enabled)
+                : null;
+        }
+
+        TransportTransaction CreateTransportTransaction(string incomingQueuePartitionKey)
+        {
+            var transportTransaction = new TransportTransaction();
+
+            if (pushSettings.RequiredTransactionMode == TransportTransactionMode.SendsAtomicWithReceive)
+            {
+                transportTransaction.Set((receiver.ServiceBusConnection, receiver.Path));
```

Comment:
Is it true that only because of this we need the ValueTuple reference for 4.6.1, see https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/pull/29/files#diff-8a7b07bd3be63680840be24e868393dfR18? If the answer is yes it is a bit overkill I think? We could just as easily build a custom struct and pass that into the transaction

Reply:
We've discussed options, including a custom struct, and decided to go with `ValueTuple` reference for 4.6.1. I suggest raise a separate issue if you feel strongly about it or have an argument that would point out the decision was not the optimal.